python/snewpy/models/base.py

```diff
@@ -0,0 +1,287 @@
+from abc import abstractmethod, ABC
+
+from astropy.io import ascii, fits
+from astropy.table import Table, join
+from astropy.units.quantity import Quantity
+
+import numpy as np
+from scipy.interpolate import interp1d
+from scipy.special import loggamma, gamma, lpmv
+
+import os
+
+from warnings import warn
+from snewpy.neutrino import Flavor
+from snewpy.flavor_transformation import *
+
+class SupernovaModel(ABC):
+    """Base class defining an interface to a supernova model."""
+    metadata = {}
+    def __init__(self):
+        pass
+
+    def __repr__(self):
+        """Default representation of the model.
+        """
+        # self.__class__ will be something like 
+        mod = f"{self.__class__.__name__} Model"
+        try:
+            mod +=f': {self.filename}'
+        except:
+            pass
+        s = [mod]
+        for name, v in self.metadata.items():
+            s +=[f"{name:16} : {v}"]
+        return '\n'.join(s)
+        
+    def _repr_markdown_(self):
+        """Markdown representation of the model, for Jupyter notebooks.
+        """
+        mod = f'**{self.__class__.__name__} Model**'
+        try:
+            mod +=f': {self.filename}'
+        except:
+            pass
+        s = [mod,'']
+        if self.metadata:
+            s += ['|Parameter|Value|',
+                  '|:--------|:----:|']
+            for name, v in self.metadata.items():
+                try: 
+                    s += [f"|{name} | ${v.value:g}$ {v.unit:latex}|"]
+                except:
+                    s += [f"|{name} | {v} |"]
+        return '\n'.join(s)
+
+    @abstractmethod
+    def get_time(self):
+        """Returns
+        -------
+            returns array of snapshot times from the simulation
+        """
+        pass
```

Comment:
Right now, this method is implemented in three different places:
* in `PinchedModel` (which most models in `snewpy.models.ccsn` inherit from; either directly or via `_GarchingArchiveModel`)
* and separately for the `Fornax_2019` and `Fornax_2021` models in `snewpy.models.ccsn`

In all three cases the implementation is simply `return self.time`. Would it make sense to move that to the base `SupernovaModel` class instead of having an abstract method here?

Reply:
I was thinking about that.
It would mean that we require each subclass to set the `self.time` attribute, which is slightly less straightforward than requiring implementing the abstract method.

I mean if one doesn't implement the abstract method, a `TypeError` is raised on construction, but if the `self.time` is not defined, we'll get `AttributeError` only at the point we try to access it.
Is there any nice way to impose this requirement?

Reply:
Well, here's one possible solution. 
I can wrap the `__init__` method of the children classes on creation, to make the checks. It can be done with Metaclass, or, slightly simpler, with [`__init_subclass__`](https://docs.python.org/3/reference/datamodel.html#object.__init_subclass__) method:
```python
def check_model(self,*arg,**kwargs):
    if hasattr(self, 'time')==False:
        raise TypeError(f'{self.__class__}.time attribute not initialized in __init__!')
    
def wrap_init(init, check):
    @wraps(init)
    def _wrapper(self, *arg, **kwargs):
        init(self,*arg,**kwargs)
        check(self)
    return _wrapper

class SupernovaModel():
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        cls.__init__ = wrap_init(cls.__init__, check_model)
```
Then having two models:
```python
class Model0(SupernovaModel):
    def __init__(self):
        self.time=[0,1,2,3]

class Model1(SupernovaModel):
    def __init__(self):
        pass
```
`Model0()` creates an instance, but `Model1()` raises `TypeError`.

Maybe it's not very important in general, but subclassing `SupernovaModel is one most possible way our code will be used. Also we can make additional checks for the models, if needed. What do you think?

Reply:
Implemented in 519d09f

Reply:
I get that it’s slightly nicer to raise an error on construction, rather than on first access; but am honestly not sure if it’s worth all this extra code. Since the target audience for this is developers who want to add their own module (who already need to read the code for the base class to figure out what functions to override), I’d be okay with letting this be a plain old AttributeError on access and just adding a comment (probably in the docstring of `SupernovaModel`’s `__init__`?) saying that the `__init__` function must set `self.time` and `self.metadata` and what their format should be.

Alternatively, I guess we could use `check_model()` to add some verification whether `self.metadata` is set & filled in?

Reply:
I'm also not sure if this is a good solution. Having abstract methods that user needs to implement seems cleaner. However if we switch to required attributes, I think it's important to have crash at the construction, otherwise user might start looking at the code in whatever downstream procedure tried to run `model.get_time`.
 
Also I agree that this is too much for a single check, and we can add more checks there.
For the `self.metadata` initialization - it was intended behavior, because I thought that in some cases no metadata can be provided by the model. But probably in that case one can set it explicitly to `{}`.

I also think the best way to make our intentions clear is to make `SupernovaModel.__init__(self, time, metadata)`which would fill the appropriate variables, so subclasses will call `super().__init__(times, metadata)` in their constructor.

Reply:
I think requiring the developer adding a model to think about metadata for a second is useful; the risk of them forgetting about it is larger than the very minor inconvenience of having to explicitly set it to `{}` in the very rare case where there actually isn’t any metadata to add.

The `super().__init__` way sounds cleaner; good idea!